service/raft/raft_group0.cc

```diff
@@ -400,19 +426,81 @@ future<> raft_group0::setup_group0(db::system_keyspace& sys_ks, const std::unord
+        }
+    }
+
+
+    group0_log.info("setup_group0: the cluster is ready to use Raft. Finishing.");
+    co_await _client.set_group0_upgrade_state(group0_upgrade_state::use_post_raft_procedures);
 }
 
-future<> raft_group0::become_voter() {
-    if (!_raft_gr.is_enabled() || !joined_group0()) {
+future<> raft_group0::finish_setup_after_join() {
+    if (joined_group0()) {
+        group0_log.info("finish_setup_after_join: group 0 ID present, loading server info.");
+        auto my_addr = co_await load_my_addr();
+        if (!_raft_gr.group0().get_configuration().can_vote(my_addr.id)) {
+            group0_log.info("finish_setup_after_join: becoming a voter in the group 0 configuration...");
+            // Just bootstrapped and joined as non-voter. Become a voter.
+            auto pause_shutdown = _shutdown_gate.hold();
+            co_await _raft_gr.group0().modify_config({{my_addr, true}}, {}, &_abort_source);
+            group0_log.info("finish_setup_after_join: became a group 0 voter.");
+
+            // No need to run `upgrade_to_group0()` since we must have bootstrapped with Raft
+            // (that's the only way to join as non-voter today).
+            co_return;
+        }
+    } else if (!_raft_gr.is_enabled()) {
+        group0_log.info("finish_setup_after_join: local RAFT feature disabled, skipping.");
         co_return;
+    } else {
+        // We're either upgrading or in recovery mode.
     }
 
-    auto my_addr = co_await load_my_addr();
-    assert(std::holds_alternative<raft::group_id>(_group0));
-    auto& gid = std::get<raft::group_id>(_group0);
-    if (!_raft_gr.get_server(gid).get_configuration().can_vote(my_addr.id)) {
-        auto pause_shutdown = _shutdown_gate.hold();
-        co_return co_await _raft_gr.group0().modify_config({{my_addr, true}}, {}, &_abort_source);
+    // Note: even if we joined group 0 before, it doesn't necessarily mean we finished with the whole
+    // upgrade procedure which has follow-up steps after joining group 0, hence we need to prepare
+    // the listener below (or fire the upgrade procedure if the feature is already enabled) even if
+    // we're already a member of group 0 by this point.
+    // `upgrade_to_group0()` will correctly detect and handle which case we're in: whether the procedure
+    // has already finished or we restarted in the middle after a crash.
+
+    if (_feat.supports_raft_cluster_mgmt) {
+        group0_log.info("finish_setup_after_join: SUPPORTS_RAFT feature enabled, starting upgrade.");
+        co_await upgrade_to_group0();
+    } else {
+        group0_log.info("finish_setup_after_join: SUPPORTS_RAFT feature not yet enabled, scheduling upgrade to start when it is.");
+        _raft_support_listener = _feat.supports_raft_cluster_mgmt.when_enabled([this] {
+            group0_log.info("finish_setup_after_join: SUPPORTS_RAFT feature enabled. Starting internal upgrade-to-raft procedure.");
+            upgrade_to_group0().get();
```

Comment:
Read `upgrade_to_group0`. The procedure already works in the background.

Reply:
Ah, you mean this part may block?
```
    auto start_state = (co_await _client.get_group0_upgrade_state()).second;
```
Indeed. So maybe the entire thing should be moved into `do_get_upgrade_to_group0`, or the entire `upgrade_to_group0` needs to be moved to background as you say...

Reply:
> Indeed.

Wait, no, it cannot block at this point.

This may take only a read lock on the upgrade state, and there is no way a write lock is taken at this point, because the write lock is taken during `set_group0_upgrade_state` which is called after the upgrade starts - and at this point we haven't started yet, just checking whether or not we should start.

Reply:
Fair enough. 

Although, it seems unsafe to rely on the implementation details of upgrade_to_group0(), which could change later without us realizing this. Better to move to background at the place which requires it, i.e the gossip handler.

Reply:
I can also expose in the interface the fact that the blocking part that depends on other nodes is running in the background  - by writing a comment at the function declaration.